cmd/podman/machine/cp.go

```diff
@@ -0,0 +1,160 @@
+	"github.com/containers/podman/v5/libpod/events"
+	"github.com/containers/podman/v5/pkg/copy"
+	"github.com/containers/podman/v5/pkg/machine/define"
+	"github.com/containers/podman/v5/pkg/machine/env"
+	"github.com/containers/podman/v5/pkg/machine/vmconfigs"
+	"github.com/containers/podman/v5/pkg/specgen"
+	"github.com/spf13/cobra"
+)
+
+type cpOptions struct {
+	Quiet    bool
+	Machine  *vmconfigs.MachineConfig
+	IsSrc    bool
+	SrcPath  string
+	DestPath string
+}
+
+var (
+	cpCmd = &cobra.Command{
+		Use:               "cp [options] SRC_PATH DEST_PATH",
+		Short:             "Securely copy contents between the virtual machine",
+		Long:              "Securely copy files or directories between the virtual machine and your host",
+		PersistentPreRunE: machinePreRunE,
+		RunE:              cp,
+		Args:              cobra.ExactArgs(2),
+		Example:           `podman machine cp ~/ca.crt podman-machine-default:/etc/containers/certs.d/ca.crt`,
+		ValidArgsFunction: autocompleteMachineCp,
+	}
+
+	cpOpts = cpOptions{}
+)
+
+func init() {
+	registry.Commands = append(registry.Commands, registry.CliCommand{
+		Command: cpCmd,
+		Parent:  machineCmd,
+	})
+
+	flags := cpCmd.Flags()
+	quietFlagName := "quiet"
+	flags.BoolVarP(&cpOpts.Quiet, quietFlagName, "q", false, "Suppress copy status output")
+}
+
+func cp(_ *cobra.Command, args []string) error {
+	var err error
+
+	srcMachine, srcPath, destMachine, destPath, err := copy.ParseSourceAndDestination(args[0], args[1])
+	if err != nil {
+		return err
+	}
+
+	// Passing an absolute windows path of the format <volume>:\<path> will cause
+	// `copy.ParseSourceAndDestination` to think the volume is a Machine. Check
+	// if the raw cmdline argument is a Windows host path.
+	if runtime.GOOS == "windows" {
+		if specgen.IsHostWinPath(args[0]) {
+			srcMachine = ""
+			srcPath = args[0]
+		}
+
+		if specgen.IsHostWinPath(args[1]) {
+			destMachine = ""
+			destPath = args[1]
+		}
+	}
+
+	mc, err := resolveMachine(srcMachine, destMachine)
+	if err != nil {
+		return err
+	}
+
+	state, err := provider.State(mc, false)
+	if err != nil {
+		return err
+	}
+	if state != define.Running {
+		return fmt.Errorf("vm %q is not running", mc.Name)
+	}
+
+	cpOpts.Machine = mc
+	cpOpts.SrcPath = srcPath
+	cpOpts.DestPath = destPath
+
+	err = secureCopy(&cpOpts)
+	if err != nil {
+		return fmt.Errorf("copy failed: %s", err.Error())
+	}
+
+	fmt.Println("Copy successful")
+	newMachineEvent(events.Copy, events.Event{Name: mc.Name})
+	return nil
+}
+
+func secureCopy(opts *cpOptions) error {
+	srcPath := opts.SrcPath
+	destPath := opts.DestPath
+	sshConfig := opts.Machine.SSH
+
+	username := sshConfig.RemoteUsername
+	if cpOpts.Machine.HostUser.Rootful {
+		username = "root"
+	}
+	username += "@localhost:"
+
+	if opts.IsSrc {
+		srcPath = username + srcPath
+	} else {
+		destPath = username + destPath
+	}
+
+	args := []string{"-r", "-i", sshConfig.IdentityPath, "-P", strconv.Itoa(sshConfig.Port),
```

Comment:
anyone else worried that we have this kind of stuff in two places now and if we decided to alter one of them we'd miss the other?

Reply:
Yes, I forgot to consolidate that. Thanks for pointing that out.

Reply:
Addressed in the latest push.